rpc/internal/collector.service.go

```diff
@@ -190,54 +205,107 @@ func (sc *streamsCoordinator) getStream(name string) (string, *streamGroup, erro
 	return id, group, nil
 }
 
-// addEvent writes the given event to the collector from the given stream. If
-// the stream does not exist an error is returned. Note that this function
-// blocks until all streams in the group have an entry in the heap, at which
-// point the timestamp can be guaranteed.
+// addEvent writes the given event to the stream's buffer. If the time since
+// the last flush has surpassed the flush interval, the flush method is called.
+// If the stream does not exist or is already closed an error is returned.
 func (sg *streamGroup) addEvent(ctx context.Context, id string, event *events.Performance) error {
 	sg.mu.Lock()
+	defer sg.mu.Unlock()
+
+	if sg.catcher.HasErrors() {
+		return sg.catcher.Resolve()
+	}
 
-	errChan, ok := sg.streams[id]
+	stream, ok := sg.streams[id]
 	if !ok {
-		sg.mu.Unlock()
-		return errors.Errorf("stream %s does not exist in this stream group", id)
+		return errors.Errorf("stream '%s' does not exist in this stream group", id)
+	}
+	if stream.closed {
+		return errors.Errorf("stream '%s' already closed", id)
 	}
 
-	sg.eventHeap.SafePush(&performanceHeapItem{errChan: errChan, event: event})
-	if sg.eventHeap.Len() >= len(sg.streams) {
-		item := sg.eventHeap.SafePop()
-		go sg.sendError(item)
+	if stream.inHeap {
+		stream.buffer <- event
+		return nil
 	}
 
-	sg.mu.Unlock()
+	var e *events.Performance
 	select {
-	case <-ctx.Done():
-		return errors.Wrap(ctx.Err(), "context canceled while adding event")
-	case err := <-errChan:
-		return err
+	case e = <-stream.buffer:
+		stream.buffer <- e
+	default:
+		e = event
 	}
-}
+	sg.eventHeap.SafePush(&performanceHeapItem{id: id, event: event})
+	stream.inHeap = true
 
-// sendError writes the next item from the heap to the collector and sends the
-// error, if any, to the corresponding error channel.
-func (sg *streamGroup) sendError(item *performanceHeapItem) {
-	if item != nil {
-		item.errChan <- sg.collector.AddEvent(item.event)
+	if time.Since(sg.lastFlush) > 5*time.Second { // TODO: figure out a good flush interval
+		return errors.Wrap(sg.flush(), "problem persisting data")
 	}
+
+	return nil
 }
 
-// removeStream removes the given stream from the stream group. If the
-// underlying min heap is full, it will pop the heap and write the item to the
-// collector.
-func (sg *streamGroup) removeStream(id string) {
+// closeStream marks the given stream as closed. Once all the items in the
+// stream's buffer have been written to the collector, the stream will be
+// removed from the stream.
+func (sg *streamGroup) closeStream(id string) {
 	sg.mu.Lock()
 	defer sg.mu.Unlock()
 
-	delete(sg.streams, id)
-	if sg.eventHeap.Len() >= len(sg.streams) {
+	stream, ok := sg.streams[id]
+	if !ok {
+		return
+	}
+	stream.closed = true
+	if len(stream.buffer) == 0 {
+		delete(sg.streams, id)
+	}
+
+	sg.catcher.Add(errors.Wrap(sg.flush(), "problem persisting data"))
```

Comment:
The only case I see in which this would actually return the errors from the catcher is in `addEvent()`. Should we also log/check the accumulated errors at some point outside of adding more events?

Reply:
Yeah it is awkward because I do a defer to close the stream, so that is why I have the catcher, but it will miss any errors from the last stream closed. Since we don't use named return values the code has to get a little messier, I will change it.

Reply:
This can be removed, we don't need the catcher. I will return an error here and that should fix it.